src/core/broker_config.cpp

```diff
@@ -17,80 +17,282 @@
+    json::value_t::number_unsigned,
+    [](const json& val, auto& config) { return ValidateAndStoreInt<unsigned int>(val, config.settings.max_connections); },
+    std::function<void(BrokerConfig&)>() // Leave the default constructed value in the settings struct
+  },
+  {
+    "/max_controllers"_json_pointer,
+    json::value_t::number_unsigned,
+    [](const json& val, auto& config) { return ValidateAndStoreInt<unsigned int>(val, config.settings.max_controllers); },
+    std::function<void(BrokerConfig&)>() // Leave the default constructed value in the settings struct
+  },
+  {
+    "/max_controller_messages"_json_pointer,
+    json::value_t::number_unsigned,
+    [](const json& val, auto& config) { return ValidateAndStoreInt<unsigned int>(val, config.settings.max_controller_messages); },
+    std::function<void(BrokerConfig&)>() // Leave the default constructed value in the settings struct
+  },
+  {
+    "/max_devices"_json_pointer,
+    json::value_t::number_unsigned,
+    [](const json& val, auto& config) { return ValidateAndStoreInt<unsigned int>(val, config.settings.max_devices); },
+    std::function<void(BrokerConfig&)>() // Leave the default constructed value in the settings struct
+  },
+  {
+    "/max_device_messages"_json_pointer,
+    json::value_t::number_unsigned,
+    [](const json& val, auto& config) { return ValidateAndStoreInt<unsigned int>(val, config.settings.max_device_messages); },
+    std::function<void(BrokerConfig&)>() // Leave the default constructed value in the settings struct
+  },
+  {
+    "/max_reject_connections"_json_pointer,
+    json::value_t::number_unsigned,
+    [](const json& val, auto& config) { return ValidateAndStoreInt<unsigned int>(val, config.settings.max_reject_connections); },
+    std::function<void(BrokerConfig&)>() // Leave the default constructed value in the settings struct
+  }
+};
+// clang-format on
+
+// Read the JSON configuration from a file.
+BrokerConfig::ParseResult BrokerConfig::Read(const std::string& file_name)
 {
   std::ifstream file_stream(file_name);
   if (!file_stream.is_open())
     return ParseResult::kFileOpenErr;
 
-  return ReadFromStream(file_stream);
+  return Read(file_stream);
 }
 
-BrokerConfig::ParseResult BrokerConfig::ReadFromStream(std::istream& stream)
+// Read the JSON configuration from an input stream.
+BrokerConfig::ParseResult BrokerConfig::Read(std::istream& stream)
 {
   try
   {
     stream >> current_;
-    return Validate(current_, defaults_);
+    return ValidateCurrent();
   }
   catch (json::parse_error)
   {
     return ParseResult::kJsonParseErr;
   }
 }
 
-BrokerConfig::ParseResult BrokerConfig::Validate(const json& current_obj, const json& default_obj)
+// Validate the JSON object contained in the "current_" member, which presumably has just been
+// deserialized. Currently, extra keys present in the JSON which we don't recognize are considered
+// valid.
+BrokerConfig::ParseResult BrokerConfig::ValidateCurrent()
```

Comment:
It feels like this function as a whole should give more useful debugging information. As a user if I make a config file that fails to validate, I want to know *why* it failed to validate. Not just mysteriously fail saying it was an invalid setting. At the least saying which key it failed on would be helpful (this could be done via logging of some sort, not necessarily adding this info to the return).

Reply:
Similarly, I don't think there's a way to check later how valid the config is. If the user handles the return codes poorly and has a problem in their config, they'll have a half-initialized BrokerConfig object laying around that could cause confusion.

Reply:
OK, yeah good idea. I might fork off another branch and PR to add a logging mechanism, since it should probably be generalized for the entire Broker anyway.

I'll add the current valid state stored as a member of the class, as well.